File: PadraoCriacionais/factoryMethod/FactoryMethod01.js
//Classe base de veiculo
class Veiculo{
    constructor(modelo){
        this.modelo = modelo;
    }
    mostrarDetalhes(){
        console.log(`modelo: ${this.modelo}`);
    }
}

//Subclasse de veiculo
class Carro extends Veiculo{
    constructor(modelo){
        super(modelo);
    }
}

class Moto extends Veiculo{
    constructor(modelo){
        super(modelo);
    }
}

class Barco extends Veiculo{
    constructor(modelo){
        super(modelo);
    }
}

//Fabrica Abstrata de veiculo
class FabricaDeVeiculos{
    criarVeiculo(modelo){
        throw new Error("O modelo criarVeiculo deve ser implementado pela subclasse")
    }
}

//Fabrica Concreta de Carros
class FabricaDeCarros extends FabricaDeVeiculos{
    criarVeiculo(modelo){
        return new Carro(modelo);
    }
}

    //Fabrica Concreta de Moto
class FabricaDeMotos extends FabricaDeVeiculos{
    criarVeiculo(modelo){
        return new Moto(modelo);
    }
}

   //Fabrica Concreta de Barco
   class FabricaDeBarcos extends FabricaDeVeiculos{
    criarVeiculo(modelo){
        return new Barco(modelo);
    }
}
//Uso de FACTORYMETHOD

const fabricaDeCarros = new FabricaDeCarros();
const carro01 = fabricaDeCarros.criarVeiculo("Hatch");
carro01.mostrarDetalhes();


const carro02 = fabricaDeCarros.criarVeiculo("Sedan");
carro02.mostrarDetalhes();

const fabricaDeMoto = new FabricaDeMotos();
const moto = fabricaDeMoto.criarVeiculo("CB");
moto.mostrarDetalhes();

const fabricaDeBarco = new FabricaDeBarcos();
const barco = fabricaDeBarco.criarVeiculo("Lancha");
barco.mostrarDetalhes();

File: PadraoCriacionais/factoryMethod/FactoryMethod01.py
#Classe base de veiculo
class Veiculo:
    def __init__(self, modelo):
        self.modelo = modelo

    def mostrarDetalhes(self):
        print(f"modelo:{self.modelo}")

#Subclasse de veiculo
class Carro(Veiculo):
    def __init__(self, modelo):
        super().__init__(modelo)   

class Moto(Veiculo):
    def __init__(self, modelo):
        super().__init__(modelo)  

class Barco(Veiculo):
    def __init__(self, modelo):
        super().__init__(modelo)            


#Fabrica Abstrata de veiculo

class FabricaDeVeiculos:
    def criarVeiculo(self, modelo):
        raise NotImplementedError("ERRO")


#Fabrica Concreta de Carros
class FabricaDeCarros (FabricaDeVeiculos):
   def criarVeiculo(self, modelo):
        return Carro(modelo)


    #Fabrica Concreta de Moto
class FabricaDeMotos (FabricaDeVeiculos):
   def criarVeiculo(self, modelo):
        return Moto(modelo)



   #Fabrica Concreta de Barco
class FabricaDeBarcos (FabricaDeVeiculos):
   def criarVeiculo(self, modelo):
        return Barco(modelo)


#Uso de FACTORYMETHOD

fabricaDeCarros = FabricaDeCarros()
carro01 = fabricaDeCarros.criarVeiculo("Hatch")
carro01.mostrarDetalhes()


carro02 = fabricaDeCarros.criarVeiculo("Sedan")
carro02.mostrarDetalhes()

fabricaDeMoto = FabricaDeMotos()
moto = fabricaDeMoto.criarVeiculo("CB")
moto.mostrarDetalhes()

fabricaDeBarco = FabricaDeBarcos()
barco = fabricaDeBarco.criarVeiculo("Lancha")
barco.mostrarDetalhes()